go/libraries/doltcore/sqle/dsess/transactions.go

```diff
@@ -112,62 +117,117 @@ func (tx *DoltTransaction) Commit(ctx *sql.Context, workingSet *doltdb.WorkingSe
+			} else if err != nil {
+				return nil, err
+			}
 
-		existingWorkingRoot := ws.WorkingRoot()
+			existingWorkingRoot := ws.WorkingRoot()
 
-		hash, err := ws.HashOf()
-		if err != nil {
-			return nil, err
-		}
+			hash, err := ws.HashOf()
+			if err != nil {
+				return nil, err
+			}
+
+			if newWorkingSet || rootsEqual(existingWorkingRoot, tx.startState.WorkingRoot()) {
+				// ff merge
+				err = tx.dbData.Ddb.UpdateWorkingSet(ctx, tx.workingSetRef, workingSet, hash, tx.getWorkingSetMeta(ctx))
+				if err == datas.ErrOptimisticLockFailed {
+					return nil, nil
+				} else if err != nil {
+					return nil, err
+				}
+
+				return workingSet, nil
+			}
+
+			start := time.Now()
+			mergedRoot, stats, err := merge.MergeRoots(ctx, existingWorkingRoot, workingSet.WorkingRoot(), tx.startState.WorkingRoot())
+			if err != nil {
+				return nil, err
+			}
+			logrus.Tracef("merge took %s", time.Since(start))
+
+			var tablesWithConflicts []string
+			for table, mergeStats := range stats {
+				if mergeStats.Conflicts > 0 {
+					if transactionMergeStomp {
+						tablesWithConflicts = append(tablesWithConflicts, table)
+					} else {
+						// TODO: surface duplicate key errors as appropriate
+						return nil, fmt.Errorf("conflict in table %s", table)
+					}
+				}
+			}
 
-		if newWorkingSet || rootsEqual(existingWorkingRoot, tx.startState.WorkingRoot()) {
-			// ff merge
-			err = tx.dbData.Ddb.UpdateWorkingSet(ctx, tx.workingSetRef, workingSet, hash, tx.getWorkingSetMeta(ctx))
+			// Only resolve conflicts automatically if the stomp environment key is set
+			if len(tablesWithConflicts) > 0 {
+				mergedRoot, err = tx.stompConflicts(ctx, mergedRoot, tablesWithConflicts)
+				if err != nil {
+					return nil, err
+				}
+			}
+
+			mergedWorkingSet := workingSet.WithWorkingRoot(mergedRoot)
+			err = tx.dbData.Ddb.UpdateWorkingSet(ctx, tx.workingSetRef, mergedWorkingSet, hash, tx.getWorkingSetMeta(ctx))
 			if err == datas.ErrOptimisticLockFailed {
-				continue
+				return nil, nil
 			} else if err != nil {
 				return nil, err
 			}
 
-			return workingSet, nil
-		}
+			return mergedWorkingSet, nil
+		}()
 
-		mergedRoot, stats, err := merge.MergeRoots(ctx, existingWorkingRoot, workingSet.WorkingRoot(), tx.startState.WorkingRoot())
 		if err != nil {
 			return nil, err
+		} else if updatedWs != nil {
+			return updatedWs, nil
 		}
+	}
 
-		for table, mergeStats := range stats {
-			if mergeStats.Conflicts > 0 {
-				// TODO: surface duplicate key errors as appropriate
-				return nil, fmt.Errorf("conflict in table %s", table)
-			}
+	// TODO: different error type for retries exhausted
+	return nil, datas.ErrOptimisticLockFailed
+}
+
+// stompConflicts resolves the conflicted tables in the root given by blindly accepting theirs, and returns the
+// updated root value
+func (tx *DoltTransaction) stompConflicts(ctx *sql.Context, mergedRoot *doltdb.RootValue, tablesWithConflicts []string) (*doltdb.RootValue, error) {
```

Comment:
is stomping conflicts the correct behavior for `REPEATABLE READ`? last write wins?

Reply:
It's defensible, but for it to be correct you need to support select for update with row-level locks.

Really the only reason to turn this on is for benchmarks, an application would want to catch this error and retry the transaction